src/System.Net.Http/src/System/Net/Http/Managed/AuthenticateAndRedirectHandler.cs

```diff
@@ -0,0 +1,223 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Net.Http.Headers;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.Http
+{
+    internal sealed partial class AuthenticateAndRedirectHandler : HttpMessageHandler
+    {
+        private readonly HttpMessageHandler _innerHandler;
+        private readonly bool _preAuthenticate;
+        private readonly ICredentials _credentials;
+        private readonly bool _allowRedirect;
+        private readonly int _maxAutomaticRedirections;
+        private AuthenticationHelper.DigestResponse _digestResponse;
+
+        public AuthenticateAndRedirectHandler(bool preAuthenticate, ICredentials credentials, bool allowRedirect, int maxAutomaticRedirections, HttpMessageHandler innerHandler)
+        {
+            Debug.Assert(innerHandler != null);
+
+            _preAuthenticate = preAuthenticate;
+            _credentials = credentials;
+            _allowRedirect = allowRedirect;
+
+            if (allowRedirect && maxAutomaticRedirections < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(maxAutomaticRedirections));
+            }
+
+            _maxAutomaticRedirections = maxAutomaticRedirections;
+            _innerHandler = innerHandler;
+        }
+
+        protected internal override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            HttpResponseMessage response;
+            uint redirectCount = 0;
+            bool useCredentialCache = false;
+            while (true)
+            {
+                response = await SendRequestAsync(request, useCredentialCache, cancellationToken).ConfigureAwait(false);
+
+                if (!RequestNeedsRedirect(response))
+                {
+                    break;
+                }
+
+                // Clear the authorization header, if the request requires redirect.
+                request.Headers.Authorization = null;
+
+                // Just as with WinHttpHandler and CurlHandler, for security reasons, we drop the server credential if it is
+                // anything other than a CredentialCache. We allow credentials in a CredentialCache since they
+                // are specifically tied to URIs.
+                useCredentialCache = true;
```

Comment:
>useCredentialCache = true; [](start = 16, length = 26)

Do we need this?  At the call to SendRequestAsync where this is consumed, won't this be true iff redirectCount > 0, in which case we can just pass redirectCount > 0 rather than tracking this separately (and consuming a field on the state machine for it)?

Reply:
Ok sure.